runtime-sdk/src/types/transaction.rs

```diff
@@ -81,6 +90,101 @@ impl UnverifiedTransaction {
     }
 }
 
+/// Transaction signer.
+pub struct TransactionSigner {
+    ut: UnverifiedTransaction,
+    tx: Transaction,
+}
+
+impl TransactionSigner {
+    /// Allocate proof structures based on the specified authentication info in the transaction.
+    fn allocate_proofs(&mut self) {
+        if !self.ut.1.is_empty() {
+            return;
+        }
+
+        // Allocate proof slots.
+        self.ut
+            .1
+            .resize_with(self.tx.auth_info.signer_info.len(), Default::default);
+
+        for (si, ap) in self
+            .tx
+            .auth_info
+            .signer_info
+            .iter()
+            .zip(self.ut.1.iter_mut())
+        {
+            match (&si.address_spec, ap) {
+                (AddressSpec::Multisig(cfg), ap) => {
+                    // Allocate multisig slots.
+                    let mut sigs = Vec::new();
+                    sigs.resize_with(cfg.signers.len(), Default::default);
+
+                    *ap = AuthProof::Multisig(sigs);
+                }
+                _ => continue,
+            }
+        }
+    }
+
+    /// Sign the transaction and append the signature.
+    ///
+    /// The signer must be specified in the `auth_info` field.
+    pub fn append_sign<S>(&mut self, signer: &S) -> Result<(), Error>
+    where
+        S: Signer + ?Sized,
+    {
+        let ctx = signature::context::get_chain_context_for(SIGNATURE_CONTEXT_BASE);
+        let signature = signer.sign(&ctx, &self.ut.0)?;
+
+        let mut matched = false;
+        self.allocate_proofs();
```

Comment:
You could also call `TransactionSigner::new(tx)` and allocate at initialization when you construct unverified transaction.